devdoc/two_dimensional_array_requirements.md

```diff
@@ -0,0 +1,106 @@
+# two_dimensional_array_requirements
+============================
+
+## Overview
+
+`two_dimensional_array` provides a 2D array with variable size of rows and columns. Each row has the same size to ensure easier access. The first dimention of array is resizable, it contains the pointer to the second dimention of array. The pointer is initiazlied to NULL and points to the allocated memory address when there's not enough space. When the memory stored in a row is not needed any more, the entire row shall be freed and set back to `NULL`.
+
+## Exposed API
+```c
+typedef struct TWO_D_ARRAY_TAG
+{
+    uint64_t rows;
+    uint64_t cols;
+    size_t item_size;
+    void * row_arrays[];
+}TWO_D_ARRAY;
+
+typedef struct TWO_D_ARRAY_TAG* TWO_D_ARRAY_HANDLE;
+MOCKABLE_FUNCTION(, TWO_D_ARRAY_HANDLE, two_d_array_create, uint64_t, row_size, uint64_t, col_size, size_t, item_size);
+MOCKABLE_FUNCTION(, int, two_d_array_free_row, TWO_D_ARRAY_HANDLE, two_d_array_handle, uint64_t, row_index);
```

Comment:
This is a free row generic API, how does this help with sp_index? I think we started at redoing sp_index....